Mat::ptr not defined for more than 3 dimensions

<!--
If you have a question rather than reporting a bug please go to http://answers.opencv.org where you get much faster responses.
If you need further assistance please read [How To Contribute](https://github.com/opencv/opencv/wiki/How_to_contribute).

This is a template helping you to create an issue which can be processed as quickly as possible. This is the bug reporting section for the OpenCV library.
-->

##### System information (version)
<!-- Example
- OpenCV => 3.1
- Operating System / Platform => Windows 64 Bit
- Compiler => Visual Studio 2015
-->

- OpenCV =>  3.3, git repository as of 9/6/17
- Operating System / Platform => probably all, tested on Linux
- Compiler => probably all C++ compilers, tested with g++ 5.4

##### Detailed description
There is no definition for the templated `ptr` method declared in `core/mat.hpp`:
 ```.cpp
template<typename _Tp> const _Tp* ptr(const int* idx) const;
```
The definition is missing in `core/mat.inl.hpp` (where it probably belongs), and the example code below generates linker's error:
    `mat4.cpp:(.text+0x113): undefined reference to `float* cv::Mat::ptr<float>(int const*)'`
<!-- your description -->

##### Steps to reproduce

Try building an executable for this source code:

```.cpp
#include <opencv2/core.hpp>
int main (int argc, const char *argv[])
{
  int dims4[4] = {2,3,4,5}, idx4[4];
  cv::Mat_<float> m4 (4, dims4, 4.5);
  idx4      [3] = 0;
  for     (int w=0; w<dims4[0]; w++) {
    idx4    [0] = w;
    for     (int z=0; z<dims4[0]; z++) {
      idx4  [1] = z;
      for   (int y=0; y<dims4[1]; y++) {
	idx4[2] = y;
	float* yline = m4.ptr<float>(idx4);
	for (int x=0; x<dims4[2]; x++)
	  yline[x] = x+y+z;
      }
    }
  }
  return 0;
}
```
<!-- to add code example fence it with triple backticks and optional file extension
    ```.cpp
    // C++ code example
    ```
 or attach as .txt or .zip file
-->

--- a/modules/core/include/opencv2/core/mat.inl.hpp
+++ b/modules/core/include/opencv2/core/mat.inl.hpp
@@ -1059,6 +1059,34 @@ const uchar* Mat::ptr(const int* idx) const
 }
 
 template<typename _Tp> inline
+_Tp* Mat::ptr(const int* idx)
+{
+    int i, d = dims;
+    uchar* p = data;
+    CV_DbgAssert( d >= 1 && p );
+    for( i = 0; i < d; i++ )
+    {
+        CV_DbgAssert( (unsigned)idx[i] < (unsigned)size.p[i] );
+        p += idx[i] * step.p[i];
+    }
+    return (_Tp*)p;
+}
+
+template<typename _Tp> inline
+const _Tp* Mat::ptr(const int* idx) const
+{
+    int i, d = dims;
+    uchar* p = data;
+    CV_DbgAssert( d >= 1 && p );
+    for( i = 0; i < d; i++ )
+    {
+        CV_DbgAssert( (unsigned)idx[i] < (unsigned)size.p[i] );
+        p += idx[i] * step.p[i];
+    }
+    return (const _Tp*)p;
+}
+
+template<typename _Tp> inline
 _Tp& Mat::at(int i0, int i1)
 {
     CV_DbgAssert(dims <= 2);

--- a/modules/core/test/test_mat.cpp
+++ b/modules/core/test/test_mat.cpp
@@ -1804,4 +1804,24 @@ TEST(Mat_, from_initializer_list)
     ASSERT_DOUBLE_EQ(norm(A, B, NORM_INF), 0.);
 }
 
+
+TEST(Mat, template_based_ptr)
+{
+    Mat mat = (Mat_<float>(2, 2) << 11.0f, 22.0f, 33.0f, 44.0f);
+    int idx[2] = {1, 0};
+    ASSERT_FLOAT_EQ(33.0f, *(mat.ptr<float>(idx)));
+    idx[0] = 1;
+    idx[1] = 1;
+    ASSERT_FLOAT_EQ(44.0f, *(mat.ptr<float>(idx)));
+}
+
+TEST(Mat_, template_based_ptr)
+{
+    int dim[4] = {2, 2, 1, 2};
+    Mat_<float> mat = (Mat_<float>(4, dim) << 11.0f, 22.0f, 33.0f, 44.0f,
+                                              55.0f, 66.0f, 77.0f, 88.0f);
+    int idx[4] = {1, 0, 0, 1};
+    ASSERT_FLOAT_EQ(66.0f, *(mat.ptr<float>(idx)));
+}
+
 #endif

